server/data/data-memory/src/main/java/org/apache/james/rrt/memory/MemoryRecipientRewriteTable.java

```diff
@@ -121,19 +122,21 @@ protected String mapAddressInternal(String user, Domain domain) throws Recipient
         if (mappingEntries.isEmpty()) {
             return null;
         }
-        Map<String, Collection<Mappings>> userMappingsMap = Multimaps.transformEntries(
-            Multimaps.index(mappingEntries, InMemoryMappingEntry::asKey),
-            (Maps.EntryTransformer<String, InMemoryMappingEntry, Mappings>)
-                (s, mappingEntry) -> MappingsImpl.fromMappings(Stream.of(mappingEntry.getMapping())))
-            .asMap();
-        return Maps.transformEntries(userMappingsMap,
-            (s, mappingsList) -> {
-                Mappings result = MappingsImpl.empty();
-                for (Mappings mappings : mappingsList) {
-                    result = result.union(mappings);
-                }
-                return result;
-            });
+        return mappingEntries.stream()
+            .collect(Guavate.toImmutableListMultimap(
+                InMemoryMappingEntry::asKey,
+                Function.identity()))
+            .asMap()
+            .entrySet()
+            .stream()
+            .map(entry -> Pair.of(entry.getKey(), toMappings(entry.getValue())))
+            .collect(Guavate.toImmutableMap(Pair::getKey, Pair::getValue));
```

Comment:
you can collect directly with Guavate.toImmutableMap(Entry::getKey, entry -> toMappings(entry.getValue)) , no?